Feature/issue 1544 include gradient and elbo exceptions

dustinvtran opened the pull request:
#### Summary:
Check for NaN's or non-finite numbers when calling methods from the model, specifically the gradient and log probability. From [stan-users](https://groups.google.com/forum/#!topic/stan-users/E7gBBpPtAoc) mailing list.

#### Intended Effect:
If either of the above happens, ADVI should throw a warning and drop that sampled gradient/ELBO evaluation and continue.

#### How to Verify:
Run models by building them, then running them, e.g. `make examples/bernoulli/bernoulli && cd examples/bernoulli && ./bernoulli variational data file=bernoulli.data.R`.

#### Side Effects:
N/A

#### Documentation:
N/A

#### Reviewer Suggestions:
N/A

syclik commented:
@dustinvtran, just looked at the CmdStan test you added. That's fine for CmdStan, but for Stan, we want to make sure things are serialized to whatever stream you were intending to stream it to. I'm trying to write a test, to help. I'll let you know when and where I put it up.

syclik commented:
yo... so the branch is completely broken; it doesn't compile (and this was before I touched it). This is why drive code with tests. 

I added a new test file at `src/test/unit/variational/advi_test.cpp` with the intention of using that to test the behavior you introduced, but I couldn't get anywhere yet. Let me know when you fix the issue and I'll help out.

dustinvtran commented:
The unit test is written now. Running the unit test without the exception throwing changes here causes it to fail; it succeeds with the changes in this pull request.

bgoodri commented:
Does it also need to check for the objective function being negative infinity (and perhaps positive infinity too)?

dustinvtran commented:
Thanks, just added them. They existed in the first commit of the pull request but then mysteriously disappeared in the second commit when I added the try/except clauses.

syclik commented:
[re: missing commits]

Are you rebasing or merging develop? That's not good if commits disappear.

On Wednesday, July 22, 2015, Dustin Tran <notifications@github.com> wrote:

> Thanks, just added them. They existed in the first commit of the pull
> request but then mysteriously disappeared in the second commit when I added
> the try/except clauses.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1545#issuecomment-123588860>.
>

dustinvtran commented:
This was initially part of another branch (which exists only locally). I was cleaning up all the code comments and following the suggestions Bob noted [in another pull request](https://github.com/stan-dev/stan/pull/1532). I realized later that the changes noted here may as wel be separate from those, so I had to manually make a new branch and add the changes.

bgoodri commented:
I think there still may be an issue in that one of my students pulled the latest version of this branch and still got

```
  250900  -802237.5              -nan            0.460

  251000  -679394.6              -nan            0.460

  251100  -597975.1              -nan            0.449

  251200  -549073.0              -nan            0.392

  251300  -621069.7              -nan            0.387

  251400  -1810373.6              -nan            0.446

  251500  -230846386.5              -nan            0.449

  251600  -747430.7              -nan            0.449

  251700  -686403.0              -nan            0.387

  251800  -716847.1              -nan            0.384

  251900  -605118.3              -nan            0.384

  252000  -565544.2              -nan            0.384

  252100  -483585.4              -nan            0.396

  252200  -659793.3              -nan            0.445

  252300  -84292366.5              -nan            0.446

  252400  -665741.1              -nan            0.446

  252500  -615233.7              -nan            0.384

  252600  -1529910891.5              -nan            0.449

  252700  -636919.9              -nan            0.460

  252800  -737601.4              -nan            0.004   MEDIAN ELBO CONVERGED
```
I don't see how `delta_ELBO_mean` could be `-nan` unless there was a `+Inf` and a `-Inf` that it tried to average over.

akucukelbir commented:
@bgoodri can you share your model and data with us? we can then pinpoint where the `nan`'s are coming from, as the source of the `nan` is not clear at the moment.

bgoodri commented:
It isn't mine, but I'll ask.

On Mon, Jul 27, 2015 at 3:33 PM, Alp Kucukelbir <notifications@github.com>
wrote:

> @bgoodri <https://github.com/bgoodri> can you share your model and data
> with us? we can then pinpoint where the nan's are coming from, as the
> source of the nan is not clear at the moment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1545#issuecomment-125319856>.
>

akucukelbir commented:
thanks in advance!

dustinvtran commented:
Thanks for sending the model/data (privately)! I ran the code over night and it still seemed to be running after 40,000 iterations:

```
Informational Message: The current gradient evaluation of the ELBO is ignored because of the following issue:
Exception thrown at line 623:
stan::math::lkj_corr_log: y is not positive definite:
1 1
1 1
If this warning occurs often then your model may be either severely ill-conditioned or misspecified.
  13600  -886053.2           106.776            0.244   MAY BE DIVERGING... INSPECT ELBO
  13700  -18380357.4           106.009            0.249   MAY BE DIVERGING... INSPECT ELBO
  13800  -514542.3           105.496            0.249   MAY BE DIVERGING... INSPECT ELBO
...
  38000  -48899540.5         12980.082            0.336   MAY BE DIVERGING... INSPECT ELBO
  38100  -1536240.3         12946.183            0.340   MAY BE DIVERGING... INSPECT ELBO
  38200  -808639.2         12912.384            0.340   MAY BE DIVERGING... INSPECT ELBO
  38300  -3024521884.3         12878.760            0.357   MAY BE DIVERGING... INSPECT ELBO
  38400  -4418484.3         12847.084            0.357   MAY BE DIVERGING... INSPECT ELBO
  38500  -1565235.1         12813.806            0.360   MAY BE DIVERGING... INSPECT ELBO
  38600  -657839.6         12780.699            0.360   MAY BE DIVERGING... INSPECT ELBO
  38700  -606121.4         12747.760            0.360   MAY BE DIVERGING... INSPECT ELBO
  38800  -538987.1         12714.989            0.357   MAY BE DIVERGING... INSPECT ELBO
  38900  -589298.8         12682.387            0.357   MAY BE DIVERGING... INSPECT ELBO
  39000  -552413.2         12649.951            0.340   MAY BE DIVERGING... INSPECT ELBO
  39100  -526179.2         12617.681            0.340   MAY BE DIVERGING... INSPECT ELBO
  39200  -2727362.8         12585.577            0.340   MAY BE DIVERGING... INSPECT ELBO
  39300  -1420503217.2         12553.637            0.357   MAY BE DIVERGING... INSPECT ELBO
  39400  -1172843.0         12524.919            0.357   MAY BE DIVERGING... INSPECT ELBO
  39500  -773067.0         12493.292            0.360   MAY BE DIVERGING... INSPECT ELBO
  39600  -12356863.2         12461.825            0.360   MAY BE DIVERGING... INSPECT ELBO
  39700  -952876.5         12430.544            0.362   MAY BE DIVERGING... INSPECT ELBO
  39800  -1167422.9         12399.390            0.360   MAY BE DIVERGING... INSPECT ELBO
  39900  -768608.9         12368.393            0.362   MAY BE DIVERGING... INSPECT ELBO
  40000  -575929.6         12337.550            0.360   MAY BE DIVERGING... INSPECT ELBO
  40100  -842209.8         12306.860            0.360   MAY BE DIVERGING... INSPECT ELBO
  40200  -20382123.6         12276.325            0.360   MAY BE DIVERGING... INSPECT ELBO
```

I haven't seen any `nan`s, as dropping the gradients as the informational message mentions was working fine for me. It's possible that the student didn't update to the latest commits on this branch? If he/she still sees it, let us know the random seed involved as well.

syclik commented:
@akucukelbir / @dustinvtran: looks like there weren't enough includes. This test is failing:
`make src/stan/variational/advi.hpp-test`

dustinvtran commented:
Here's the error I got from the compilation:
```{bash}
dvt ~/github/cmdstan/stan make src/stan/variational/advi.hpp-test
g++ -I src -isystem lib/eigen_3.2.4 -isystem lib/boost_1.58.0 -isystem lib/stan_math_2.7.0 -Wall -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -pipe -Wno-unused-function -ftemplate-depth-256   -c -O0 -include src/stan/variational/advi.hpp -o /dev/null test/dummy.cpp
In file included from <built-in>:326:
In file included from <command line>:4:
In file included from ./src/stan/variational/advi.hpp:4:
In file included from lib/stan_math_2.7.0/stan/math/prim/scal/err/check_finite.hpp:8:
lib/stan_math_2.7.0/stan/math/prim/scal/fun/value_of_rec.hpp:25:14: error: cannot convert 'const Eigen::Matrix<double, -1, 1, 0, -1, 1>' to 'double' without a conversion operator
      return static_cast<double>(x);
             ^~~~~~~~~~~~~~~~~~~~~~
lib/stan_math_2.7.0/stan/math/prim/scal/err/check_not_nan.hpp:36:38: note: in instantiation of function template specialization 'stan::math::value_of_rec<Eigen::Matrix<double, -1,
      1, 0, -1, 1> >' requested here
            if ((boost::math::isnan)(value_of_rec(stan::get(y, n))))
                                     ^
lib/stan_math_2.7.0/stan/math/prim/scal/err/check_not_nan.hpp:67:11: note: in instantiation of member function 'stan::math::(anonymous namespace)::not_nan<Eigen::Matrix<double, -1,
      1, 0, -1, 1>, true>::check' requested here
        ::check(function, name, y);
          ^
src/stan/variational/families/normal_fullrank.hpp:74:21: note: in instantiation of function template specialization 'stan::math::check_not_nan<Eigen::Matrix<double, -1, 1, 0, -1, 1>
      >' requested here
        stan::math::check_not_nan(function, "Mean vector", mu_);
                    ^
In file included from <built-in>:326:
In file included from <command line>:4:
In file included from ./src/stan/variational/advi.hpp:4:
In file included from lib/stan_math_2.7.0/stan/math/prim/scal/err/check_finite.hpp:8:
lib/stan_math_2.7.0/stan/math/prim/scal/fun/value_of_rec.hpp:25:14: error: cannot convert 'const Eigen::Matrix<double, -1, -1, 0, -1, -1>' to 'double' without a conversion operator
      return static_cast<double>(x);
             ^~~~~~~~~~~~~~~~~~~~~~
lib/stan_math_2.7.0/stan/math/prim/scal/err/check_not_nan.hpp:36:38: note: in instantiation of function template specialization 'stan::math::value_of_rec<Eigen::Matrix<double, -1,
      -1, 0, -1, -1> >' requested here
            if ((boost::math::isnan)(value_of_rec(stan::get(y, n))))
                                     ^
lib/stan_math_2.7.0/stan/math/prim/scal/err/check_not_nan.hpp:67:11: note: in instantiation of member function 'stan::math::(anonymous namespace)::not_nan<Eigen::Matrix<double, -1,
      -1, 0, -1, -1>, true>::check' requested here
        ::check(function, name, y);
          ^
src/stan/variational/families/normal_fullrank.hpp:75:21: note: in instantiation of function template specialization 'stan::math::check_not_nan<Eigen::Matrix<double, -1, -1, 0, -1,
      -1> >' requested here
        stan::math::check_not_nan(function, "Cholesky factor", L_chol_);
                    ^
2 errors generated.
make: *** [src/stan/variational/advi.hpp-test] Error 1
```
Not sure how to fix that.  The call `stan::math::check_not_nan(function, "Mean vector", mu_);` checks if the `Eigen::VectorXd` object `mu_` is not `nan`, and looking at the `check_not_nan` function, it looks like it should work—and it does when running ADVI. I'm not familiar enough to know what's going on here for these local compilations(?).

syclik commented:
fixed it -- just needed to include `stan/math.hpp`. We have some template traits that need to be included in the right order.

bob-carpenter commented:
We need to get this one done and in before we can put ADVI through to the interfaces.

akucukelbir commented:
the travis tests seem to have passed, but not jenkins? i'm a bit confused.

syclik commented:
Nothing to be confused about. They're all different configurations. This
one is failing under Windows. See:
http://d1m1s1b1.stat.columbia.edu:8080/job/Stan%20Pull%20Request%20-%20Windows%20-%20Tests%20-%20Unit/329/warnings18Result/new/

You might want to add a 'u' at the end of the number.
http://stackoverflow.com/questions/2347936/cant-get-rid-of-this-decimal-constant-is-unsigned-only-in-iso-c90-warning


On Tue, Jul 28, 2015 at 9:07 PM, Alp Kucukelbir <notifications@github.com>
wrote:

> the travis tests seem to have passed, but not jenkins? i'm a bit confused.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1545#issuecomment-125794847>.
>

akucukelbir commented:
thanks daniel. this passes now. feel free to merge, if all looks good on your end.

syclik commented on src/stan/variational/advi.hpp:
```diff
@@ -94,15 +89,30 @@ namespace stan {
+            elbo += energy_i;
+          } catch (std::exception& e) {
+            this->write_error_msg_(print_stream_, e);
+            i -= 1;
```
instead of this, which takes a minute to process what you're doing, could you do something like:
```cpp
n = 0;
int n_monte_carlo_drop = 0;
while (n < n_monte_carlo_elbo_ && n_monte_carlo_drop < n_monte_carlo_elbo_) {
...
}
```

Or something like that. Decrementing the loop index within the loop is just tricky to figure out.

bob-carpenter replied:
Ack --- Dustin's already changed this, which is why I'm not seeing
what Daniel was objecting to any more.  I think the rewrite's going
in the right direction.  I had a few more concerns:

* check_finite should catch NaN, so I don't think you also need check_not_nan

* variables should be declared as close to where they're used as possible,
so the energy_i variable should be down in the loop with

  double energy_i
    = model_.template ...

* the zeta is an exception because you don't want to have to create
new vectors.  That's why you probably want to pass the vector
zeta as a non-constant reference argument to variational.sample(rng_, zeta)
so that the sample() method can fill it --- otherwise you'll be creating
and discarding vectors anyway and having to do a loop of the dimensionality
for assignment


        double elbo = 0.0;
        double energy_i;
        int dim = variational.dimension();
        Eigen::VectorXd zeta(dim);

        int i = 0;
        int n_monte_carlo_drop = 0;
        while (i < n_monte_carlo_elbo_) {
          zeta = variational.sample(rng_);
          try {
            energy_i = model_.template log_prob<false, true>(zeta,
                                                             print_stream_);
            stan::math::check_not_nan(function, "energy_i", energy_i);
            stan::math::check_finite(function, "energy_i", energy_i);
            elbo += energy_i;
            i += 1;
          } catch (std::exception& e) {
            this->write_error_msg_(print_stream_, e);
            n_monte_carlo_drop += 1;
            stan::math::check_less(function, "n_monte_carlo_drop",
              n_monte_carlo_drop, n_monte_carlo_elbo_);
          }
        }

- Bob



> On Jul 29, 2015, at 5:12 PM, Daniel Lee <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> >  
> >          for (int i = 0; i < n_monte_carlo_elbo_; ++i) {
> >            zeta = variational.sample(rng_);
> > -          // Accumulate log probability
> > -          elbo += model_.template log_prob<false, true>(zeta, print_stream_);
> > +          try {
> > +            energy_i = model_.template log_prob<false, true>(zeta,
> > +                                                             print_stream_);
> > +            stan::math::check_not_nan(function, "energy_i", energy_i);
> > +            stan::math::check_finite(function, "energy_i", energy_i);
> > +            elbo += energy_i;
> > +          } catch (std::exception& e) {
> > +            this->write_error_msg_(print_stream_, e);
> > +            i -= 1;
> 
> instead of this, which takes a minute to process what you're doing, could you do something like:
> 
> n = 0;
> int n_monte_carlo_drop = 0;
> while (n < n_monte_carlo_elbo_ && n_monte_carlo_drop < n_monte_carlo_elbo_) {
> ...
> }
> 
> Or something like that. Decrementing the loop index within the loop is just tricky to figure out.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgoodri commented:
Due to some recent commit, it is now possible to get this message
```
std::exception::what: stan::variational::normal_meanfield::calc_grad: n_monte_carlo_drop is 1, but must be less than 1
```
I assume this means something like all proposals got rejected, but I think we need a more expansive error message here telling people what to do.

dustinvtran commented:
I took out `stan::math::check_less` and essentially wrote my own version in order to have the more verbose error message while calling `stan::math::domain_error`. Not sure if there's a more elegant way to do this. The error message is now
```
stan::variational::normal_meanfield::calc_grad: The number of dropped evaluations has reached its maximum amount (500). Your model may be either severely ill-conditioned or misspecified.
```
In practice, the user would see this after already seeing (in this case) 500 individual warnings saying that a gradient was dropped.

dustinvtran commented:
Not sure what's going on. Looks like the warnings are unrelated to anything here? http://d1m1s1b1.stat.columbia.edu:8080/job/Stan%20Pull%20Request%20-%20Tests%20-%20Unit/503/warnings18Result/HIGH/

syclik commented:
I'll hunt this down. I've been seeing this a bit... not sure what causes it
yet.

On Tue, Aug 4, 2015 at 1:13 PM, Dustin Tran <notifications@github.com>
wrote:

> Not sure what's going on. Looks like the warnings are unrelated to
> anything here?
> http://d1m1s1b1.stat.columbia.edu:8080/job/Stan%20Pull%20Request%20-%20Tests%20-%20Unit/503/warnings18Result/HIGH/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1545#issuecomment-127679054>.
>

syclik commented:
Jenkins, retest this please.

syclik commented:
@dustinvtran, looks good.

dustinvtran commented:
:dancer:
